File: backend/app/schemas/analytics.py
"""
Analytics schemas for request/response validation
"""
from pydantic import BaseModel, field_serializer
from typing import List, Optional
from datetime import datetime
from uuid import UUID


class AnalyticsQueryParams(BaseModel):
    """Query parameters for analytics endpoints"""
    time_range: Optional[str] = "all_time"  # last_24h, last_7d, last_30d, all_time
    start_date: Optional[datetime] = None
    end_date: Optional[datetime] = None
    project_ids: Optional[List[str]] = None


class AnalyticsSummaryResponse(BaseModel):
    """Aggregated summary metrics"""
    total_traces: int
    total_llm_calls: int
    total_tool_calls: int
    total_input_tokens: int
    total_output_tokens: int
    total_tokens: int
    total_cost: float
    avg_duration_ms: float
    min_duration_ms: float
    max_duration_ms: float
    total_duration_ms: float
    unique_projects: int


class TrendDataPoint(BaseModel):
    """Single data point in time series"""
    timestamp: datetime
    input_tokens: int
    output_tokens: int
    total_tokens: int
    cost: float
    trace_count: int


class TrendsResponse(BaseModel):
    """Time series data with granularity"""
    data: List[TrendDataPoint]
    granularity: str  # "hour", "day", "week"


class ModelBreakdownItem(BaseModel):
    """Cost and usage breakdown for a single model"""
    model_name: str
    provider: str
    total_cost: float
    cost_percentage: float
    input_tokens: int
    output_tokens: int
    total_tokens: int
    call_count: int


class ModelsResponse(BaseModel):
    """List of model breakdowns"""
    models: List[ModelBreakdownItem]


class TopTraceItem(BaseModel):
    """High-usage trace information"""
    trace_id: str
    name: str
    total_tokens: int
    total_cost: float
    duration_ms: float
    llm_call_count: int
    start_time: datetime
    project_name: str
    status: str


class TopTracesResponse(BaseModel):
    """List of top traces"""
    traces: List[TopTraceItem]

File: frontend/types/analytics.ts
export interface AnalyticsSummary {
  total_traces: number
  total_llm_calls: number
  total_tool_calls: number
  total_input_tokens: number
  total_output_tokens: number
  total_tokens: number
  total_cost: number
  avg_duration_ms: number
  min_duration_ms: number
  max_duration_ms: number
  total_duration_ms: number
  unique_projects: number
}

export interface TrendData {
  timestamp: string
  input_tokens: number
  output_tokens: number
  total_tokens: number
  cost: number
  trace_count: number
}

export interface TrendsResponse {
  data: TrendData[]
  granularity: string
}

export interface ModelBreakdown {
  model_name: string
  provider: string
  total_cost: number
  cost_percentage: number
  input_tokens: number
  output_tokens: number
  total_tokens: number
  call_count: number
}

export interface TopTrace {
  trace_id: string
  name: string
  total_tokens: number
  total_cost: number
  duration_ms: number
  llm_call_count: number
  start_time: string
  project_name: string
  status: string
}


export interface AnalyticsFilters {
  timeRange: 'last_24h' | 'last_7d' | 'last_30d' | 'all_time' | 'custom'
  customStartDate: Date | null
  customEndDate: Date | null
  selectedProjects: string[]
}
